Add container registry cache on repository host.

nuclearsandwich opened the pull request:
This PR adds a configured a pull-through Docker registry cache on the repository host.
The cache is configured by default but is not used by any ros_buildfarm nodes unless a corresponding address is added to the `node['docker']['registry_mirrors']` attribute.

There is another PR coming which adds a customized seccomp profile that will almost certainly conflict with this one. It doesn't really matter which one lands first but let's go in order of first-opened-first-merged and I'll rebase that PR once this one has merged.

nuclearsandwich committed:
Add container registry cache on repository host.

nuclearsandwich committed:
Add docker daemon configuration for registry mirrors.

j-rivero commented on recipes/agent.rb:
```diff
@@ -1,5 +1,15 @@
 apt_update
 package 'docker.io'
+service 'docker' do
```
can the service be added after the template to save the `notifies` or is there any benefit in starting it first and restart just after deploying the configuration?

nuclearsandwich replied:
In general yes, I try to order recipes with package, config files, service.

Due to the development order I didn't do that here and should. But we can't remove the notification because we want to make sure that a change to the config file triggers a restart if the docker service is already running.

nuclearsandwich replied:
Services have been moved below configuration in the latest patch.

j-rivero commented on recipes/repo.rb:
```diff
@@ -272,3 +272,15 @@
+  package 'docker-registry'
+  cookbook_file '/etc/docker/registry/config.yml' do
+    source 'docker-registry-config.yml'
+    notifies :restart, 'service[docker-registry]'
```
Do this restart just before the service do anything different than just remove it?

nuclearsandwich replied:
notification only occurs when the file is changed. So when chef is run and this file doesn't change, it won't restart the service. But if the file does change and the service is always running, without the notification to restart the docker-registry service wouldn't pick up the new config.

j-rivero replied:
thanks for the explanation.

j-rivero commented on templates/docker-daemon.json.erb:
```diff
@@ -0,0 +1,4 @@
+{
+  <%= if not @registry_mirrors.empty? then %("registry-mirrors": [ ) + @registry_mirrors.map{|m| %("#{m}")}.join(", ") + " ],\n" end -%>
+  "seccomp-profile": "/etc/docker/seccomp/default-with-personality.json"
```
This file is going to be added in an upcoming PR, right?

nuclearsandwich replied:
Right, I didn't mean to commit this.

nuclearsandwich replied:
It's been rebased out.

j-rivero reviewed (commented):
Looks good, minor comments

nuclearsandwich committed:
Move service resource after configuration files.

In general I aim for

1. install package
2. place configuration files
3. activate service

but I messed that up here.

j-rivero commented:
Looks good to me. It would be nice to add a docker operation (might be a single pull) in the tests so we can check that the cache is not heavily broken.

nuclearsandwich committed:
Add tests for docker-registry cache usage.

nuclearsandwich committed:
Fix trailing comma in json template.

nuclearsandwich commented:
> It would be nice to add a docker operation (might be a single pull) in the tests so we can check that the cache is not heavily broken.

As has been discussed before, I do not see the appeal of a lot of the sample infrastructure tests which just express the same information as the configuration but as constraints rather than declarations. However, at your prompting I added tests that verify the output of `docker info` and `docker run hello-world` and in fact caught a bug introduced by splitting the seccomp profile changes out of this PR. However, we can't run docker inside the dokken runners on GitHub Actions so I'll have to figure out how to mask out those tests.

nuclearsandwich committed:
These tests caught a bug but can't be run in kitchen-dokken.

nuclearsandwich commented:
>  I'll have to figure out how to mask out those tests.

I couldn't figure out a way to detect running within kitchen-dokken in short order so I've commented those tests out entirely for now.

j-rivero commented:
> I couldn't figure out a way to detect running within kitchen-dokken in short order so I've commented those tests out entirely for now.

Disabling it if we detect that code is running on a docker container (checking for `/.dockerenv` file) will disable it in more occasions that it should but will improve the situation where the code is always disabled.

nuclearsandwich commented:
> Disabling it if we detect that code is running on a docker container (checking for `/.dockerenv` file) will disable it in more occasions that it should but will improve the situation where the code is always disabled.

That's a good idea. I'll keep it in mind as a way to bring those tests back in the future.
